### PR TITLE
Always cache Attribute compatibility and disambiguation rules

### PR DESCRIPTION
**Pull request by jvandort**
Compatibility rules especially are called in tight loops in the middle of dependency resolution. We have seen builds that rely on custom rules heavily spend a huge amount of time during resolution instantiating these rules over and over.

This commit updates the rule instantiator to cache the value for the rule so that we do not waste time instantiating the rule over and over again.

We expect these rules to be deterministic. Both AGP and KGP rules were verified to be deterministic and not have side-effects in their constructors.

Fixes [#30110](https://github.com/gradle/gradle/issues/30110)

Running the Android 2000 project extra large benchmark with 10 iterations:

before:

Max: 71.0 seconds
Min: 60.5 seconds
Mean: 63.0 seconds
Median: 62.3 seconds

after:

Max: 63.4 seconds
Min: 53.8 seconds
Mean: 56.4 seconds
Median: 55.4 seconds

This gives an improvement of:

Max: 10.8%
Min: 11.1%
Mean: 10.5%
Median: 11.1%

This is clear looking a the flamegraphs of artifact selection for example
image
image

Artifact selection went from 10,842 samples to 6,646 samples, a 38.7% improvement

The raw flamegraphs themselves:

[androidStudioSync-8 11-20240809025650+0000-cpu-simplified-flames](https://github.com/user-attachments/assets/ad35a34e-b70b-426a-bed0-2dc54931ccdc)
[androidStudioSync-8 11-20240809022902+0000-cpu-simplified-flames](https://github.com/user-attachments/assets/a93c77c4-ddfe-40f0-ae69-8abfe1ea3792)

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test apt

**Comment by jvandort** [comment]
@bot-gradle test apt

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/ResolvedVariantSet.java [hunk 1]
```diff
@@ -44,7 +44,7 @@ public interface ResolvedVariantSet {
     /**
      * The variants available for artifact selection when variant reselection is not enabled.
      */
-    Set<ResolvedVariant> getVariants();
+    List<ResolvedVariant> getVariants();
```
Comment (big-guy) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/ResolvedVariantSet.java:47]: Why are you changing this? Is this just to get rid of the `as Set` in the tests?  List implies that you could have multiple equivalent `ResolvedVariant`s

Some alternatives to that would be:
- use `toSet()`
- declare the type as a Set vs def

`as Set` is idiomatic Groovy, so I wouldn't really worry about that.

Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/ResolvedVariantSet.java:47]: It showed up in the flameraphs

Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/ResolvedVariantSet.java:47]: The set operations, implemented by hashing, are not free

**Comment by jvandort** [comment]
We will not always cache these since we want to enforce statelessness of the rules. 

Instad, we will:
1. Use https://github.com/gradle/gradle/blob/9239189ef8d307b43b765f5b9637f509d8f5a436/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/InstantiationScheme.java#L38 to improve instantiation performance
2. Cache results of each rule call
3. Hoist up attribute schema caches so we deduplicate caches between projects

**Pull request closed without merging**